server/src/main/java/org/elasticsearch/index/shard/IndexLongFieldRange.java

```diff
@@ -335,4 +342,30 @@ public int hashCode() {
         result = 31 * result + Arrays.hashCode(shards);
         return result;
     }
+
+    /**
+     * Remove the given shard from the set of known shards, possibly without adjusting the min and max. Used when allocating a stale primary
+     * which may have a different range from the original, so we must allow the range to grow. Note that this doesn't usually allow the
+     * range to shrink, so we may in theory hit this shard more than needed after allocating a stale primary.
+     */
+    public IndexLongFieldRange removeShard(int shardId, int numberOfShards) {
+        assert 0 <= shardId && shardId < numberOfShards : shardId + " vs " + numberOfShards;
+
+        if (shards != null && Arrays.stream(shards).noneMatch(i -> i == shardId)) {
+            return this;
+        }
+        if (this == MUTABLE) {
```

Comment:
I think we could allow going from MUTABLE to UNKNOWN in the single shard case, i.e., move this below the if-statement below?

Reply:
True, addressed in https://github.com/elastic/elasticsearch/pull/65564/commits/5bc657e3a069cbb97683897cb41308c36ce80cbe.